src/shims/unix/fs.rs

```diff
@@ -127,6 +129,89 @@ impl FileDescription for FileHandle {
+            let host_op = match op {
+                SharedLock { nonblocking: false } => libc::LOCK_SH,
+                SharedLock { nonblocking: true } => libc::LOCK_SH | libc::LOCK_NB,
+                ExclusiveLock { nonblocking: false } => libc::LOCK_EX,
+                ExclusiveLock { nonblocking: true } => libc::LOCK_EX | libc::LOCK_NB,
+                Unlock => libc::LOCK_UN,
+            };
+
+            let fd = self.file.as_raw_fd();
+            let ret = unsafe { libc::flock(fd, host_op) };
+            let res = match ret {
+                0 => Ok(()),
+                -1 => Err(io::Error::last_os_error()),
+                ret => panic!("Unexpected return value from flock: {ret}"),
+            };
+            Ok(res)
+        }
+
+        #[cfg(target_family = "windows")]
+        {
+            use std::os::windows::io::AsRawHandle;
+            use windows_sys::Win32::{
+                Foundation::{ERROR_IO_PENDING, ERROR_LOCK_VIOLATION, FALSE, HANDLE, TRUE},
+                Storage::FileSystem::{
+                    LockFileEx, UnlockFile, LOCKFILE_EXCLUSIVE_LOCK, LOCKFILE_FAIL_IMMEDIATELY,
+                },
+            };
+            let handle = self.file.as_raw_handle() as HANDLE;
+
+            use FlockOp::*;
+            let mut lock_nb = false;
+            let ret = match op {
+                SharedLock { nonblocking } | ExclusiveLock { nonblocking } => {
+                    lock_nb = nonblocking;
+                    let mut flags = 0;
+                    if nonblocking {
+                        flags |= LOCKFILE_FAIL_IMMEDIATELY;
+                    }
+                    if let ExclusiveLock { .. } = op {
+                        flags |= LOCKFILE_EXCLUSIVE_LOCK;
+                    }
+                    unsafe { LockFileEx(handle, flags, 0, !0, !0, &mut std::mem::zeroed()) }
+                }
+                Unlock => unsafe { UnlockFile(handle, 0, 0, !0, !0) },
+            };
+
+            let res = match ret {
+                TRUE => Ok(()),
+                FALSE => {
+                    let mut err = io::Error::last_os_error();
+                    let code: u32 = err.raw_os_error().unwrap().try_into().unwrap();
+                    // Replace error with a custom WouldBlock error, which later will be mapped
+                    // in the `helpers` module
+                    if lock_nb && matches!(code, ERROR_IO_PENDING | ERROR_LOCK_VIOLATION) {
+                        let desc = format!("LockFileEx wouldblock error: {err}");
+                        err = io::Error::new(io::ErrorKind::WouldBlock, desc);
+                    }
+                    Err(err)
+                }
+                _ => panic!("Unexpected return value: {ret}"),
+            };
+            Ok(res)
+        }
+
+        #[cfg(not(any(target_family = "unix", target_family = "windows")))]
+        {
+            let _ = op;
+            throw_unsup_format!("flock is supported only on UNIX and Windows hosts");
```

Comment:
Do we even have other hosts? We could probably just `compile_error!` here

Reply:
We have a bunch of code with `cfg(unix)` and `cfg(windows)` and nothing else. Miri does currently not build on other hosts.

Reply:
Lately, I prefer to use `target_family` instead of `cfg(unix/windows)`. I think having this future-proofing is fine.